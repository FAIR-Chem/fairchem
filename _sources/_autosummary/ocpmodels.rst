ocpmodels
=========

.. automodule:: ocpmodels

   
   
   

   
   
   

   
   
   

   
   
   



.. rubric:: Modules

.. autosummary::
   :toctree:
   :recursive:

   ocpmodels.common
   ocpmodels.datasets
   ocpmodels.models
   ocpmodels.modules
   ocpmodels.preprocessing
   ocpmodels.tasks
   ocpmodels.trainers

