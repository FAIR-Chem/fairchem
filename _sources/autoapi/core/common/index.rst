:py:mod:`core.common`
=====================

.. py:module:: core.common

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   relaxation/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   data_parallel/index.rst
   distutils/index.rst
   flags/index.rst
   gp_utils/index.rst
   hpo_utils/index.rst
   logger/index.rst
   registry/index.rst
   transforms/index.rst
   tutorial_utils/index.rst
   typing/index.rst
   utils/index.rst


