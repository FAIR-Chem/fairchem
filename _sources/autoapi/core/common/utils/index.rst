core.common.utils
=================

.. py:module:: core.common.utils

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Classes
-------

.. autoapisummary::

   core.common.utils.UniqueKeyLoader
   core.common.utils.Complete
   core.common.utils.SeverityLevelBetween


Functions
---------

.. autoapisummary::

   core.common.utils.pyg2_data_transform
   core.common.utils.save_checkpoint
   core.common.utils.warmup_lr_lambda
   core.common.utils.print_cuda_usage
   core.common.utils.conditional_grad
   core.common.utils.plot_histogram
   core.common.utils.collate
   core.common.utils.add_edge_distance_to_graph
   core.common.utils._import_local_file
   core.common.utils.setup_experimental_imports
   core.common.utils._get_project_root
   core.common.utils.setup_imports
   core.common.utils.dict_set_recursively
   core.common.utils.parse_value
   core.common.utils.create_dict_from_args
   core.common.utils.load_config
   core.common.utils.build_config
   core.common.utils.create_grid
   core.common.utils.save_experiment_log
   core.common.utils.get_pbc_distances
   core.common.utils.radius_graph_pbc
   core.common.utils.get_max_neighbors_mask
   core.common.utils.get_pruned_edge_idx
   core.common.utils.merge_dicts
   core.common.utils.setup_logging
   core.common.utils.compute_neighbors
   core.common.utils.check_traj_files
   core.common.utils.new_trainer_context
   core.common.utils._resolve_scale_factor_submodule
   core.common.utils._report_incompat_keys
   core.common.utils.load_state_dict
   core.common.utils.scatter_det
   core.common.utils.get_commit_hash
   core.common.utils.cg_change_mat
   core.common.utils.irreps_sum
   core.common.utils.update_config
   core.common.utils.get_loss_module


Module Contents
---------------

.. py:class:: UniqueKeyLoader(stream)

   Bases: :py:obj:`yaml.SafeLoader`


   .. py:method:: construct_mapping(node, deep=False)


.. py:function:: pyg2_data_transform(data: torch_geometric.data.Data)

   if we're on the new pyg (2.0 or later) and if the Data stored is in older format
   we need to convert the data to the new format


.. py:function:: save_checkpoint(state, checkpoint_dir: str = 'checkpoints/', checkpoint_file: str = 'checkpoint.pt') -> str

.. py:class:: Complete

   .. py:method:: __call__(data)


.. py:function:: warmup_lr_lambda(current_step: int, optim_config)

   Returns a learning rate multiplier.
   Till `warmup_steps`, learning rate linearly increases to `initial_lr`,
   and then gets multiplied by `lr_gamma` every time a milestone is crossed.


.. py:function:: print_cuda_usage() -> None

.. py:function:: conditional_grad(dec)

   Decorator to enable/disable grad depending on whether force/energy predictions are being made


.. py:function:: plot_histogram(data, xlabel: str = '', ylabel: str = '', title: str = '')

.. py:function:: collate(data_list)

.. py:function:: add_edge_distance_to_graph(batch, device='cpu', dmin: float = 0.0, dmax: float = 6.0, num_gaussians: int = 50)

.. py:function:: _import_local_file(path: pathlib.Path, *, project_root: pathlib.Path) -> None

   Imports a Python file as a module

   :param path: The path to the file to import
   :type path: Path
   :param project_root: The root directory of the project (i.e., the "ocp" folder)
   :type project_root: Path


.. py:function:: setup_experimental_imports(project_root: pathlib.Path) -> None

   Import selected directories of modules from the "experimental" subdirectory.

   If a file named ".include" is present in the "experimental" subdirectory,
   this will be read as a list of experimental subdirectories whose module
   (including in any subsubdirectories) should be imported.

   :param project_root: The root directory of the project (i.e., the "ocp" folder)


.. py:function:: _get_project_root() -> pathlib.Path

   Gets the root folder of the project (the "ocp" folder)
   :return: The absolute path to the project root.


.. py:function:: setup_imports(config: dict | None = None) -> None

.. py:function:: dict_set_recursively(dictionary, key_sequence, val) -> None

.. py:function:: parse_value(value)

   Parse string as Python literal if possible and fallback to string.


.. py:function:: create_dict_from_args(args: list, sep: str = '.')

   Create a (nested) dictionary from console arguments.
   Keys in different dictionary levels are separated by sep.


.. py:function:: load_config(path: str, previous_includes: list | None = None)

.. py:function:: build_config(args, args_override)

.. py:function:: create_grid(base_config, sweep_file: str)

.. py:function:: save_experiment_log(args, jobs, configs)

.. py:function:: get_pbc_distances(pos, edge_index, cell, cell_offsets, neighbors, return_offsets: bool = False, return_distance_vec: bool = False)

.. py:function:: radius_graph_pbc(data, radius, max_num_neighbors_threshold, enforce_max_neighbors_strictly: bool = False, pbc=None)

.. py:function:: get_max_neighbors_mask(natoms, index, atom_distance, max_num_neighbors_threshold, degeneracy_tolerance: float = 0.01, enforce_max_strictly: bool = False)

   Give a mask that filters out edges so that each atom has at most
   `max_num_neighbors_threshold` neighbors.
   Assumes that `index` is sorted.

   Enforcing the max strictly can force the arbitrary choice between
   degenerate edges. This can lead to undesired behaviors; for
   example, bulk formation energies which are not invariant to
   unit cell choice.

   A degeneracy tolerance can help prevent sudden changes in edge
   existence from small changes in atom position, for example,
   rounding errors, slab relaxation, temperature, etc.


.. py:function:: get_pruned_edge_idx(edge_index, num_atoms: int, max_neigh: float = 1000000000.0) -> torch.Tensor

.. py:function:: merge_dicts(dict1: dict, dict2: dict)

   Recursively merge two dictionaries.
   Values in dict2 override values in dict1. If dict1 and dict2 contain a dictionary as a
   value, this will call itself recursively to merge these dictionaries.
   This does not modify the input dictionaries (creates an internal copy).
   Additionally returns a list of detected duplicates.
   Adapted from https://github.com/TUM-DAML/seml/blob/master/seml/utils.py

   :param dict1: First dict.
   :type dict1: dict
   :param dict2: Second dict. Values in dict2 will override values from dict1 in case they share the same key.
   :type dict2: dict

   :returns: **return_dict** -- Merged dictionaries.
   :rtype: dict


.. py:class:: SeverityLevelBetween(min_level: int, max_level: int)

   Bases: :py:obj:`logging.Filter`


   Filter instances are used to perform arbitrary filtering of LogRecords.

   Loggers and Handlers can optionally use Filter instances to filter
   records as desired. The base filter class only allows events which are
   below a certain point in the logger hierarchy. For example, a filter
   initialized with "A.B" will allow events logged by loggers "A.B",
   "A.B.C", "A.B.C.D", "A.B.D" etc. but not "A.BB", "B.A.B" etc. If
   initialized with the empty string, all events are passed.


   .. py:method:: filter(record) -> bool

      Determine if the specified record is to be logged.

      Returns True if the record should be logged, or False otherwise.
      If deemed appropriate, the record may be modified in-place.



.. py:function:: setup_logging() -> None

.. py:function:: compute_neighbors(data, edge_index)

.. py:function:: check_traj_files(batch, traj_dir) -> bool

.. py:function:: new_trainer_context(*, config: dict[str, Any], distributed: bool = False)

.. py:function:: _resolve_scale_factor_submodule(model: torch.nn.Module, name: str)

.. py:function:: _report_incompat_keys(model: torch.nn.Module, keys: torch.nn.modules.module._IncompatibleKeys, strict: bool = False) -> tuple[list[str], list[str]]

.. py:function:: load_state_dict(module: torch.nn.Module, state_dict: collections.abc.Mapping[str, torch.Tensor], strict: bool = True) -> tuple[list[str], list[str]]

.. py:function:: scatter_det(*args, **kwargs)

.. py:function:: get_commit_hash()

.. py:function:: cg_change_mat(ang_mom: int, device: str = 'cpu') -> torch.tensor

.. py:function:: irreps_sum(ang_mom: int) -> int

   Returns the sum of the dimensions of the irreps up to the specified angular momentum.

   :param ang_mom: max angular momenttum to sum up dimensions of irreps


.. py:function:: update_config(base_config)

   Configs created prior to FAIRChem/OCP 2.0 are organized a little different than they
   are now. Update old configs to fit the new expected structure.


.. py:function:: get_loss_module(loss_name)

