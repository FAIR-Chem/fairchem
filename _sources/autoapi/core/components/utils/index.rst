core.components.utils
=====================

.. py:module:: core.components.utils

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Classes
-------

.. autoapisummary::

   core.components.utils.ManagedAttribute


Module Contents
---------------

.. py:class:: ManagedAttribute(enforced_type: type | None = None)

   A descriptor helper to manage setting/access to an attribute of a class


   .. py:attribute:: _enforced_type


   .. py:method:: __set_name__(owner: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, name: str)


   .. py:method:: __get__(obj: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, objtype: type[fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer] | None = None)


   .. py:method:: __set__(obj: fairchem.core.components.runner.Runner | fairchem.core.components.reducer.Reducer, value: Any)


