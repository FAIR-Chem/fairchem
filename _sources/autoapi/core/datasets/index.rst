core.datasets
=============

.. py:module:: core.datasets


Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/datasets/embeddings/index


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/datasets/_utils/index
   /autoapi/core/datasets/ase_datasets/index
   /autoapi/core/datasets/base_dataset/index
   /autoapi/core/datasets/lmdb_database/index
   /autoapi/core/datasets/lmdb_dataset/index
   /autoapi/core/datasets/oc22_lmdb_dataset/index
   /autoapi/core/datasets/target_metadata_guesser/index


Classes
-------

.. autoapisummary::

   core.datasets.AseDBDataset
   core.datasets.AseReadDataset
   core.datasets.AseReadMultiStructureDataset
   core.datasets.LMDBDatabase
   core.datasets.LmdbDataset


Functions
---------

.. autoapisummary::

   core.datasets.create_dataset
   core.datasets.data_list_collater


Package Contents
----------------

.. py:class:: AseDBDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset connects to an ASE Database, allowing the storage of atoms objects
   with a variety of backends including JSON, SQLite, and database server options.

   For more information, see:
   https://databases.fysik.dtu.dk/ase/ase/db/db.html

   :param config:
                  src (str): Either
                          - the path an ASE DB,
                          - the connection address of an ASE DB,
                          - a folder with multiple ASE DBs,
                          - a list of folders with ASE DBs
                          - a glob string to use to find ASE DBs, or
                          - a list of ASE db paths/addresses.
                          If a folder, every file will be attempted as an ASE DB, and warnings
                          are raised for any files that can't connect cleanly

                          Note that for large datasets, ID loading can be slow and there can be many
                          ids, so it's advised to make loading the id list as easy as possible. There is not
                          an obvious way to get a full list of ids from most ASE dbs besides simply looping
                          through the entire dataset. See the AseLMDBDataset which was written with this usecase
                          in mind.

                  connect_args (dict): Keyword arguments for ase.db.connect()

                  select_args (dict): Keyword arguments for ase.db.select()
                          You can use this to query/filter your database

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the database

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transforms (dict[str, dict]): Dictionary specifying data transforms as {transform_function: config}
                          where config is a dictionary specifying arguments to the transform_function

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional
   :param transform: deprecated?
   :type transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config: dict) -> list[int]


   .. py:method:: get_atoms(idx: int) -> ase.Atoms

      Get atoms object corresponding to datapoint idx. Useful to read other properties not in data object.
      :param idx: index in dataset
      :type idx: int

      :returns: ASE atoms corresponding to datapoint idx
      :rtype: atoms



   .. py:method:: connect_db(address: str | pathlib.Path, connect_args: dict | None = None) -> ase.db.core.Database
      :staticmethod:



   .. py:method:: __del__()


   .. py:method:: sample_property_metadata(num_samples: int = 100) -> dict


   .. py:method:: get_relaxed_energy(identifier)
      :abstractmethod:



.. py:class:: AseReadDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset uses ase.io.read to load data from a directory on disk.
   This is intended for small-scale testing and demonstrations of OCP.
   Larger datasets are better served by the efficiency of other dataset types
   such as LMDB.

   For a full list of ASE-readable filetypes, see
   https://wiki.fysik.dtu.dk/ase/ase/io/io.html

   :param config: src (str): The source folder that contains your ASE-readable files

                  pattern (str): Filepath matching each file you want to read
                          ex. "*/POSCAR", "*.cif", "*.xyz"
                          search recursively with two wildcards: "**/POSCAR" or "**/*.cif"

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the files you read (ex. OUTCAR)

                  ase_read_args (dict): Keyword arguments for ase.io.read()

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  include_relaxed_energy (bool): Include the relaxed energy in the resulting data object.
                          The relaxed structure is assumed to be the final structure in the file
                          (e.g. the last frame of a .traj).

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transform_args (dict): Additional keyword arguments for the transform callable

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config) -> list[pathlib.Path]


   .. py:method:: get_atoms(idx: str | int) -> ase.Atoms


   .. py:method:: get_relaxed_energy(identifier) -> float


.. py:class:: AseReadMultiStructureDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset can read multiple structures from each file using ase.io.read.
   The disadvantage is that all files must be read at startup.
   This is a significant cost for large datasets.

   This is intended for small-scale testing and demonstrations of OCP.
   Larger datasets are better served by the efficiency of other dataset types
   such as LMDB.

   For a full list of ASE-readable filetypes, see
   https://wiki.fysik.dtu.dk/ase/ase/io/io.html

   :param config: src (str): The source folder that contains your ASE-readable files

                  pattern (str): Filepath matching each file you want to read
                          ex. "*.traj", "*.xyz"
                          search recursively with two wildcards: "**/POSCAR" or "**/*.cif"

                  index_file (str): Filepath to an indexing file, which contains each filename
                          and the number of structures contained in each file. For instance:

                          /path/to/relaxation1.traj 200
                          /path/to/relaxation2.traj 150

                          This will overrule the src and pattern that you specify!

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the files you read (ex. OUTCAR)

                  ase_read_args (dict): Keyword arguments for ase.io.read()

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  include_relaxed_energy (bool): Include the relaxed energy in the resulting data object.
                          The relaxed structure is assumed to be the final structure in the file
                          (e.g. the last frame of a .traj).

                  use_tqdm (bool): Use TQDM progress bar when initializing dataset

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transform_args (dict): Additional keyword arguments for the transform callable

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional
   :param transform: Additional preprocessing function for the Data object
   :type transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config) -> list[str]


   .. py:method:: get_atoms(idx: str) -> ase.Atoms


   .. py:method:: sample_property_metadata(num_samples: int = 100) -> dict


   .. py:method:: get_relaxed_energy(identifier) -> float


.. py:function:: create_dataset(config: dict[str, Any], split: str) -> Subset

   Create a dataset from a config dictionary

   :param config: dataset config dictionary
   :type config: dict
   :param split: name of split
   :type split: str

   :returns: dataset subset class
   :rtype: Subset


.. py:class:: LMDBDatabase(filename: str | pathlib.Path | None = None, create_indices: bool = True, use_lock_file: bool = False, serial: bool = False, readonly: bool = False, *args, **kwargs)

   Bases: :py:obj:`ase.db.core.Database`


   Base class for all databases.


   .. py:attribute:: readonly


   .. py:attribute:: ids
      :value: []



   .. py:attribute:: deleted_ids
      :value: []



   .. py:method:: __enter__() -> typing_extensions.Self


   .. py:method:: __exit__(exc_type, exc_value, tb) -> None


   .. py:method:: close() -> None


   .. py:method:: _write(atoms: ase.Atoms | ase.db.row.AtomsRow, key_value_pairs: dict, data: dict | None, idx: int | None = None) -> None


   .. py:method:: _update(idx: int, key_value_pairs: dict | None = None, data: dict | None = None)


   .. py:method:: _write_deleted_ids()


   .. py:method:: delete(ids: list[int]) -> None

      Delete rows.



   .. py:method:: _get_row(idx: int, include_data: bool = True)


   .. py:method:: _get_row_by_index(index: int, include_data: bool = True)

      Auxiliary function to get the ith entry, rather than a specific id



   .. py:method:: _select(keys, cmps: list[tuple[str, str, str]], explain: bool = False, verbosity: int = 0, limit: int | None = None, offset: int = 0, sort: str | None = None, include_data: bool = True, columns: str = 'all')


   .. py:property:: metadata
      Load the metadata from the DB if present


   .. py:property:: _nextid
      Get the id of the next row to be written


   .. py:method:: count(selection=None, **kwargs) -> int

      Count rows.

      See the select() method for the selection syntax.  Use db.count() or
      len(db) to count all rows.



   .. py:method:: _load_ids() -> None

      Load ids from the DB

      Since ASE db ids are mostly 1-N integers, but can be missing entries
      if ids have been deleted. To save space and operating under the assumption
      that there will probably not be many deletions in most OCP datasets,
      we just store the deleted ids.



.. py:class:: LmdbDataset(config)

   Bases: :py:obj:`fairchem.core.datasets.base_dataset.BaseDataset`


   Base Dataset class for all OCP datasets.


   .. py:attribute:: sharded
      :type:  bool

      Dataset class to load from LMDB files containing relaxation
      trajectories or single point computations.
      Useful for Structure to Energy & Force (S2EF), Initial State to
      Relaxed State (IS2RS), and Initial State to Relaxed Energy (IS2RE) tasks.
      The keys in the LMDB must be integers (stored as ascii objects) starting
      from 0 through the length of the LMDB. For historical reasons any key named
      "length" is ignored since that was used to infer length of many lmdbs in the same
      folder, but lmdb lengths are now calculated directly from the number of keys.
      :param config: Dataset configuration
      :type config: dict


   .. py:attribute:: path


   .. py:attribute:: key_mapping


   .. py:attribute:: transforms


   .. py:method:: __getitem__(idx: int) -> T_co


   .. py:method:: connect_db(lmdb_path: pathlib.Path | None = None) -> lmdb.Environment


   .. py:method:: __del__()


   .. py:method:: sample_property_metadata(num_samples: int = 100)


.. py:function:: data_list_collater(data_list: list[torch_geometric.data.data.BaseData], otf_graph: bool = False, to_dict: bool = False) -> torch_geometric.data.data.BaseData | dict[str, torch.Tensor]

