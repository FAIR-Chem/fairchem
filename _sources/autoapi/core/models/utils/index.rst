:py:mod:`core.models.utils`
===========================

.. py:module:: core.models.utils


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   activations/index.rst
   basis/index.rst


