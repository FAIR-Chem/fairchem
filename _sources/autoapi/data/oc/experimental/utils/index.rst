:py:mod:`data.oc.experimental.utils`
====================================

.. py:module:: data.oc.experimental.utils


Module Contents
---------------


Functions
~~~~~~~~~

.. autoapisummary::

   data.oc.experimental.utils.v0_check
   data.oc.experimental.utils.restart_bug_check
   data.oc.experimental.utils.plot_traj



.. py:function:: v0_check(full_traj, initial)

   Checks whether the initial structure as gathered from the POSCAR input file
   is in agreement with the initial image of the full trajectory. If not, the
   trajectory comes fro the V0 dataset which failed to save intermediate
   checkpoints.

   Args
   full_traj (list of Atoms objects): Calculated full trajectory.
   initial (Atoms object): Starting image provided by POSCAR..


.. py:function:: restart_bug_check(full_traj)

   Observed that some of the trajectories had a strange identically cyclical
   behavior - suggesting that a checkpoint was restarted from an earlier
   checkpoint rather than the latest. Checks whether the trajectory provided
   falls within that bug.

   Args
   full_traj (list of Atoms objects): Calculated full trajectory.


.. py:function:: plot_traj(traj, fname)

   Plots the energy profile of a given trajectory

   Args
   traj (list of Atoms objects): Full trajectory to be plotted
   fname (str): Filename to be used as title and save figure as.


