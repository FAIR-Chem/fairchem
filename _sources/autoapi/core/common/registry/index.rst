:py:mod:`core.common.registry`
==============================

.. py:module:: core.common.registry

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.

   # Copyright (c) Meta, Inc. and its affiliates.
   # Borrowed from https://github.com/facebookresearch/pythia/blob/master/pythia/common/registry.py.

   Registry is central source of truth. Inspired from Redux's concept of
   global store, Registry maintains mappings of various information to unique
   keys. Special functions in registry can be used as decorators to register
   different kind of classes.

   Import the global registry object using

   ``from fairchem.core.common.registry import registry``

   Various decorators for registry different kind of classes with unique keys

   - Register a model: ``@registry.register_model``



Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   core.common.registry.Registry



Functions
~~~~~~~~~

.. autoapisummary::

   core.common.registry._get_absolute_mapping



Attributes
~~~~~~~~~~

.. autoapisummary::

   core.common.registry.R
   core.common.registry.NestedDict
   core.common.registry.registry


.. py:data:: R

   

.. py:data:: NestedDict

   

.. py:function:: _get_absolute_mapping(name: str)


.. py:class:: Registry


   Class for registry object which acts as central source of truth.

   .. py:attribute:: mapping
      :type: ClassVar[NestedDict]

      

   .. py:method:: register_task(name: str)
      :classmethod:

      Register a new task to registry with key 'name'
      :param name: Key with which the task will be registered.

      Usage::
          from fairchem.core.common.registry import registry
          from fairchem.core.tasks import BaseTask
          @registry.register_task("train")
          class TrainTask(BaseTask):
              ...


   .. py:method:: register_dataset(name: str)
      :classmethod:

      Register a dataset to registry with key 'name'

      :param name: Key with which the dataset will be registered.

      Usage::

          from fairchem.core.common.registry import registry
          from fairchem.core.datasets import BaseDataset

          @registry.register_dataset("qm9")
          class QM9(BaseDataset):
              ...


   .. py:method:: register_model(name: str)
      :classmethod:

      Register a model to registry with key 'name'

      :param name: Key with which the model will be registered.

      Usage::

          from fairchem.core.common.registry import registry
          from fairchem.core.modules.layers import CGCNNConv

          @registry.register_model("cgcnn")
          class CGCNN():
              ...


   .. py:method:: register_logger(name: str)
      :classmethod:

      Register a logger to registry with key 'name'

      :param name: Key with which the logger will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          @registry.register_logger("wandb")
          class WandBLogger():
              ...


   .. py:method:: register_trainer(name: str)
      :classmethod:

      Register a trainer to registry with key 'name'

      :param name: Key with which the trainer will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          @registry.register_trainer("active_discovery")
          class ActiveDiscoveryTrainer():
              ...


   .. py:method:: register(name: str, obj) -> None
      :classmethod:

      Register an item to registry with key 'name'

      :param name: Key with which the item will be registered.

      Usage::

          from fairchem.core.common.registry import registry

          registry.register("config", {})


   .. py:method:: __import_error(name: str, mapping_name: str) -> RuntimeError
      :classmethod:


   .. py:method:: get_class(name: str, mapping_name: str)
      :classmethod:


   .. py:method:: get_task_class(name: str)
      :classmethod:


   .. py:method:: get_dataset_class(name: str)
      :classmethod:


   .. py:method:: get_model_class(name: str)
      :classmethod:


   .. py:method:: get_logger_class(name: str)
      :classmethod:


   .. py:method:: get_trainer_class(name: str)
      :classmethod:


   .. py:method:: get(name: str, default=None, no_warning: bool = False)
      :classmethod:

      Get an item from registry with key 'name'

      :param name: Key whose value needs to be retrieved.
      :type name: string
      :param default: If passed and key is not in registry, default value will
                      be returned with a warning. Default: None
      :param no_warning: If passed as True, warning when key doesn't exist
                         will not be generated. Useful for cgcnn's
                         internal operations. Default: False
      :type no_warning: bool

      Usage::

          from fairchem.core.common.registry import registry

          config = registry.get("config")


   .. py:method:: unregister(name: str)
      :classmethod:

      Remove an item from registry with key 'name'

      :param name: Key which needs to be removed.

      Usage::

          from fairchem.core.common.registry import registry

          config = registry.unregister("config")



.. py:data:: registry

   

