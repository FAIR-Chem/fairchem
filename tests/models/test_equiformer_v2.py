"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

import io
import os

import pytest
import requests
import torch
from ase.io import read

from ocpmodels.common.registry import registry
from ocpmodels.common.utils import load_state_dict, setup_imports
from ocpmodels.datasets import data_list_collater
from ocpmodels.preprocessing import AtomsToGraphs


@pytest.fixture(scope="class")
def load_data(request):
    atoms = read(
        os.path.join(os.path.dirname(os.path.abspath(__file__)), "atoms.json"),
        index=0,
        format="json",
    )
    a2g = AtomsToGraphs(
        max_neigh=200,
        radius=6,
        r_edges=False,
        r_fixed=True,
    )
    data_list = a2g.convert_all([atoms])
    request.cls.data = data_list[0]


@pytest.fixture(scope="class")
def load_model(request):
    torch.manual_seed(4)
    setup_imports()

    # download and load weights.
    checkpoint_url = "https://dl.fbaipublicfiles.com/opencatalystproject/models/2023_06/oc20/s2ef/eq2_31M_ec4_allmd.pt"

    # load buffer into memory as a stream
    # and then load it with torch.load
    r = requests.get(checkpoint_url, stream=True)
    r.raise_for_status()
    checkpoint = torch.load(
        io.BytesIO(r.content), map_location=torch.device("cpu")
    )

    model = registry.get_model_class("equiformer_v2")(
        None,
        -1,
        1,
        use_pbc=True,
        regress_forces=True,
        otf_graph=True,
        max_neighbors=20,
        max_radius=12.0,
        max_num_elements=90,
        num_layers=8,
        sphere_channels=128,
        attn_hidden_channels=64,
        num_heads=8,
        attn_alpha_channels=64,
        attn_value_channels=16,
        ffn_hidden_channels=128,
        norm_type="layer_norm_sh",
        lmax_list=[4],
        mmax_list=[2],
        grid_resolution=18,
        num_sphere_samples=128,
        edge_channels=128,
        use_atom_edge_embedding=True,
        distance_function="gaussian",
        num_distance_basis=512,
        attn_activation="silu",
        use_s2_act_attn=False,
        ffn_activation="silu",
        use_gate_act=False,
        use_grid_mlp=True,
        alpha_drop=0.1,
        drop_path_rate=0.1,
        proj_drop=0.0,
        weight_init="uniform",
    )

    new_dict = {
        k[len("module.") * 2 :]: v for k, v in checkpoint["state_dict"].items()
    }
    load_state_dict(model, new_dict)

    # Precision errors between mac vs. linux compound with multiple layers,
    # so we explicitly set the number of layers to 1 (instead of all 8).
    # The other alternative is to have different snapshots for mac vs. linux.
    model.num_layers = 1
    request.cls.model = model


@pytest.mark.usefixtures("load_data")
@pytest.mark.usefixtures("load_model")
class TestEquiformerV2:
    def test_energy_force_shape(self, snapshot):
        # Recreate the Data object to only keep the necessary features.
        data = self.data

        # Pass it through the model.
        energy, forces = self.model(data_list_collater([data]))

        assert snapshot == energy.shape
        assert snapshot == pytest.approx(energy.detach())

        assert snapshot == forces.shape
        assert snapshot == pytest.approx(forces.detach().mean(0))
