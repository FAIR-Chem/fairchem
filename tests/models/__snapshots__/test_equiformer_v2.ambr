# name: TestEquiformerV2.test_energy_force_shape
  Size(
    1,
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.1
  Approx(
  	array([0.12408737], dtype=float32)
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.2
  Size(
    34,
    3,
  )
# ---
# name: TestEquiformerV2.test_energy_force_shape.3
  Approx(
  	array([ 1.4928582e-03, -7.4169380e-05,  2.9909369e-03], dtype=float32)
  )
# ---
